sdks/python/apache_beam/runners/common.py

```diff
@@ -168,22 +169,41 @@ def __init__(self, obj_to_invoke, method_name):
     self.has_userstate_arguments = False
     self.state_args_to_replace = {}
     self.timer_args_to_replace = {}
+    self.timestamp_arg_name = None
+    self.window_arg_name = None
+
     for kw, v in zip(args[-len(defaults):], defaults):
       if isinstance(v, core.DoFn.StateParam):
         self.state_args_to_replace[kw] = v.state_spec
         self.has_userstate_arguments = True
       elif isinstance(v, core.DoFn.TimerParam):
         self.timer_args_to_replace[kw] = v.timer_spec
         self.has_userstate_arguments = True
+      elif isinstance(v, core._DoFnParam) and \
```

Comment:
Should we elevate them to their own classes similar to StateParam and TimerParam and also modify core.py (https://github.com/apache/beam/blob/8154ec5bfd74f58a9a0317792dc8274702a6cf25/sdks/python/apache_beam/transforms/core.py#L430) a little bit, so that the linked comment correctly reflects what could be passed to timer/state methods.

Reply:
Or at the very least could this be simplified with `v == core.DoFn.TimestampParam`?

Reply:
Also, possibly not in this PR, but it'd be nice to remove the redundancy between this and the PerWindowInvoker below.

Reply:
@aaltay that is a good idea. I am thinking of addressing refactor as a separate PR. are you ok with this?